Sources/GroundState/py_variational_montecarlo.hpp

```diff
@@ -29,9 +29,84 @@ namespace py = pybind11;
 
 namespace netket {
 
+namespace detail {
+template <class T, int ExtraFlags>
+py::array_t<T, ExtraFlags> as_readonly(py::array_t<T, ExtraFlags> array) {
+  py::detail::array_proxy(array.ptr())->flags &=
+      ~py::detail::npy_api::NPY_ARRAY_WRITEABLE_;
+  return array;
+}
+}  // namespace detail
+
 void AddVariationalMonteCarloModule(py::module &m) {
   auto m_vmc = m.def_submodule("variational");
 
+  py::class_<Stats>(m_vmc, "Stats")
+      .def_readonly("mean", &Stats::mean)
+      .def_readonly("error_of_mean", &Stats::error_of_mean)
+      .def_readonly("variance", &Stats::variance)
+      .def_readonly("autocorrelation", &Stats::correlation)
+      .def_readonly("R", &Stats::R);
+
+  py::class_<MCResult>(m_vmc, "MCResult",
+                       R"EOF(Result of Monte Carlo sampling.)EOF")
+      .def_property_readonly(
+          "samples",
+          [](const MCResult &self) {
+            assert(self.samples.rows() % self.n_chains == 0);
+            return detail::as_readonly(py::array_t<double, py::array::c_style>{
+                {self.samples.rows() / self.n_chains, self.n_chains,
+                 self.samples.cols()},
+                self.samples.data(),
+                py::none()});
+          },
+          py::keep_alive<1, 0>{},
+          R"EOF(Visible configurations `{vᵢ}` visited during sampling.
+
+                Visible configurations are represented by a row-major matrix
+                of `float64` where every row is a visible configuration.)EOF")
+      .def_property_readonly(
+          "log_values",
+          [](const MCResult &self) {
+            assert(self.log_values.rows() % self.n_chains == 0);
+            return detail::as_readonly(py::array_t<Complex, py::array::c_style>{
+                {self.log_values.rows() / self.n_chains, self.n_chains},
+                self.log_values.data(),
+                py::none()});
+          },
+          py::keep_alive<1, 0>{},
+          R"EOF(A vector of `complex128` representing `Ψ(vᵢ)` for all
+                    sampled visible configurations `vᵢ`.)EOF")
+      .def_property_readonly(
+          "der_logs",
+          [](const MCResult &self) -> py::object {
+            if (self.der_logs.has_value()) {
+              assert(self.der_logs->rows() % self.n_chains == 0);
+              return detail::as_readonly(
+                  py::array_t<Complex, py::array::c_style>{
+                      {self.der_logs->rows() / self.n_chains, self.n_chains,
+                       self.der_logs->cols()},
+                      self.der_logs->data(),
+                      py::none()});
+            }
+            return py::none();
+          },
+          py::keep_alive<1, 0>{},
+          R"EOF(A matrix of logarithmic derivatives.
+
+                Each row in the matrix corresponds to the gradient of
+                `Ψ(vᵢ)` with respect to variational parameters.)EOF")
+      .def_readonly("n_chains", &MCResult::n_chains,
```

Comment:
Why is there `MCResult.n_chains` but not `n_samples`? Since as far as I understand `result.samples.shape == n_samples, n_chains, _`, we should probably expose both or none of them.